nexus/src/app/silo.rs

```diff
@@ -6,30 +6,49 @@
 
 use crate::authz::ApiResource;
 use crate::context::OpContext;
-use crate::db;
 use crate::db::identity::{Asset, Resource};
 use crate::db::lookup::LookupPath;
 use crate::db::model::Name;
 use crate::db::model::SshKey;
+use crate::db::{self, lookup};
 use crate::external_api::params;
 use crate::external_api::shared;
 use crate::{authn, authz};
 use anyhow::Context;
 use nexus_db_model::UserProvisionType;
 use omicron_common::api::external::http_pagination::PaginatedBy;
-use omicron_common::api::external::DeleteResult;
 use omicron_common::api::external::Error;
 use omicron_common::api::external::ListResultVec;
 use omicron_common::api::external::LookupResult;
 use omicron_common::api::external::UpdateResult;
 use omicron_common::api::external::{CreateResult, LookupType};
 use omicron_common::api::external::{DataPageParams, ResourceType};
+use omicron_common::api::external::{DeleteResult, NameOrId};
 use omicron_common::bail_unless;
+use ref_cast::RefCast;
 use std::str::FromStr;
 use uuid::Uuid;
 
 impl super::Nexus {
     // Silos
+    pub fn silo_lookup<'a>(
+        &'a self,
+        opctx: &'a OpContext,
+        silo: &'a NameOrId,
+    ) -> LookupResult<lookup::Silo<'a>> {
+        match silo {
+            NameOrId::Id(id) => {
+                let silo =
+                    LookupPath::new(opctx, &self.db_datastore).silo_id(*id);
+                Ok(silo)
+            }
+            NameOrId::Name(name) => {
+                let silo = LookupPath::new(opctx, &self.db_datastore)
+                    .silo_name(Name::ref_cast(name));
+                Ok(silo)
+            }
+        }
+    }
```

Comment:
This lookup mechanism differs from the normal lookups in that it doesn't take a selector. Once we've moved over to the V1 API I fully expect we'll simplify the selector implementations anyway so I think it's fine that this one case is slightly different.

Reply:
I'm doing something similar here:
https://github.com/oxidecomputer/omicron/pull/2455/files#diff-882861bfb859d6d0519f20dfcb9abad82e23d52a6963f921fb144a8d9fc2f273R810